controllers/helmrelease_controller_release.go

```diff
@@ -0,0 +1,232 @@
+		return ctrl.Result{}, err
+	}
+
+	// Check chart readiness
+	if hc.Generation != hc.Status.ObservedGeneration || !meta.HasReadyCondition(hc.Status.Conditions) {
+		err := fmt.Errorf("HelmChart '%s/%s' is not ready", hc.GetNamespace(), hc.GetName())
+		return ctrl.Result{}, err
+	}
+
+	// Compose the values for this release, and record the checksum
+	values, err := r.composeValues(ctx, *hr)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	valuesChecksum := util.ValuesChecksum(values)
+
+	// Retrieve the last release for observation
+	rel, err := run.ObserveLastRelease(*hr)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If new state, mark release as progressing which resets the conditions
+	if hr.Status.LastAttemptedRevision != hc.GetArtifact().Revision ||
+		hr.Status.LastAttemptedValuesChecksum != valuesChecksum ||
+		hr.Status.LastReleaseRevision != util.ReleaseRevision(rel) ||
+		hr.Status.ObservedGeneration != hr.Generation {
+		v2.HelmReleaseProgressing(hr)
+		if err := r.Client.Status().Update(ctx, hr); err != nil {
+			return ctrl.Result{}, err
+		}
+		r.recordReadiness(hr, false)
+	}
+
+	// If the previous release was successful, return
+	if released := meta.GetCondition(hr.Status.Conditions, v2.ReleasedCondition); released != nil {
+		if released.Status == v1.ConditionTrue {
+			return ctrl.Result{}, nil
+		}
+	}
+
+	// Check if the previous remediation attempt failed
+	if remediated := meta.GetCondition(hr.Status.Conditions, v2.RemediatedCondition); remediated != nil {
+		if remediated.Status == v1.ConditionFalse {
+			err = fmt.Errorf("previous remediation attempt failed")
+			return ctrl.Result{}, err
+		}
+		switch rel {
+		case nil:
+			if hr.Spec.GetInstall().GetRemediation().RetriesExhausted(hr) {
+				err = fmt.Errorf("install retries exhausted")
+				return ctrl.Result{}, err
+			}
+		default:
+			if hr.Spec.GetUpgrade().GetRemediation().RetriesExhausted(hr) {
+				err = fmt.Errorf("upgrade retries exhausted")
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	// Load the chart
+	loadedChart, err := r.loadHelmChart(hc)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Record the release attempt
+	v2.HelmReleaseAttempted(hr, hc.GetArtifact().Revision, valuesChecksum)
+
+	// Perform the release
+	doRelease := run.Install
+	remediation := hr.Spec.GetInstall().GetRemediation()
+	successReason, failureReason := v2.InstallSucceededReason, v2.InstallFailedReason
+	if rel != nil {
+		doRelease = run.Upgrade
+		remediation = hr.Spec.GetUpgrade().GetRemediation()
+		successReason, failureReason = v2.UpgradeSucceededReason, v2.UpgradeFailedReason
+	}
+	rel, err = doRelease(*hr, loadedChart, values)
+	if err != nil {
+		v2.SetHelmReleaseCondition(hr, v2.ReleasedCondition, v1.ConditionFalse, failureReason, "")
+		remediation.IncrementFailureCount(hr)
+		r.event(*hr, hr.Status.LastAttemptedRevision, events.EventSeverityError,
+			fmt.Sprintf("Helm release failed: %s", err.Error()))
+		return ctrl.Result{}, err
+	}
+
```

Comment:
I think we still need to account for deleting the `RemediatedCondition` here:

https://github.com/fluxcd/helm-controller/blob/9b0daa031acfad05e148f0fae76cdaa454b4a658/controllers/helmrelease_controller.go#L393-L396